File: relayer/metrics.go
package relayer

import (
	"context"
	"fmt"
	"math"
	"net/http"
	"time"

	sdklogging "github.com/Layr-Labs/eigensdk-go/logging"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
)

const RelayerNamespace = "sffl_relayer"

type EventListener interface {
	OnBlockReceived()
	OnDaSubmissionFailed()
	OnDaSubmitted(duration time.Duration)
	OnRetriesRequired(retries int)
	OnInvalidNonce()
}

type SelectiveListener struct {
	OnBlockReceivedCb      func()
	OnDaSubmissionFailedCb func()
	OnDaSubmittedCb        func(duration time.Duration)
	OnRetriesRequiredCb    func(retries int)
	OnInvalidNonceCb       func()
}

func (l *SelectiveListener) OnBlockReceived() {
	if l.OnBlockReceivedCb != nil {
		l.OnBlockReceivedCb()
	}
}

func (l *SelectiveListener) OnDaSubmissionFailed() {
	if l.OnDaSubmissionFailedCb != nil {
		l.OnDaSubmissionFailedCb()
	}
}

func (l *SelectiveListener) OnDaSubmitted(duration time.Duration) {
	if l.OnDaSubmittedCb != nil {
		l.OnDaSubmittedCb(duration)
	}
}

func (l *SelectiveListener) OnRetriesRequired(retries int) {
	if l.OnRetriesRequiredCb != nil {
		l.OnRetriesRequiredCb(retries)
	}
}

func (l *SelectiveListener) OnInvalidNonce() {
	if l.OnInvalidNonceCb != nil {
		l.OnInvalidNonceCb()
	}
}

func MakeRelayerMetrics(registry *prometheus.Registry) (EventListener, error) {
	numBlocksReceived := prometheus.NewCounter(
		prometheus.CounterOpts{
			Namespace: RelayerNamespace,
			Name:      "num_blocks_received",
			Help:      "The number of blocks received from rollup",
		})

	if err := registry.Register(numBlocksReceived); err != nil {
		return nil, fmt.Errorf("error registering numBlocksReceived counter: %w", err)
	}

	numDaSubmissionsFailed := prometheus.NewCounter(prometheus.CounterOpts{
		Namespace: RelayerNamespace,
		Name:      "num_da_submissions_failed",
		Help:      "The number of failed da submissions",
	})

	if err := registry.Register(numDaSubmissionsFailed); err != nil {
		return nil, fmt.Errorf("error registering numDaSubmissionsFailed counter: %w", err)
	}

	latencyBuckets := []float64{
		1,
		25,
		50,
		75,
		100,
		250,
		500,
		1000, // 1 sec
		2000,
		3000,
		4000,
		5000,
		10000,
		math.Inf(0),
	}

	submissionDuration := prometheus.NewHistogram(prometheus.HistogramOpts{
		Namespace: RelayerNamespace,
		Name:      "submission_duration_ms",
		Help:      "Duration of successful DA submissions",
		Buckets:   latencyBuckets,
	})

	if err := registry.Register(submissionDuration); err != nil {
		return nil, fmt.Errorf("error registering submissionDuration histogram: %w", err)
	}

	retriesHistogram := prometheus.NewHistogram(prometheus.HistogramOpts{
		Namespace: RelayerNamespace,
		Name:      "retries_histogram",
		Help:      "Histogram of retry counts",
		Buckets:   prometheus.LinearBuckets(1, 1, SUBMIT_BLOCK_RETRIES),
	})

	if err := registry.Register(retriesHistogram); err != nil {
		return nil, fmt.Errorf("error registering retriesHistogram histogram: %w", err)
	}

	numInvalidNonces := prometheus.NewCounter(prometheus.CounterOpts{
		Namespace: RelayerNamespace,
		Name:      "num_of_invalid_nonces",
		Help:      "Number of InvalidNonce error",
	})
	if err := registry.Register(numInvalidNonces); err != nil {
		return nil, fmt.Errorf("error registering numInvalidNonces count: %w", err)
	}

	return &SelectiveListener{
		OnBlockReceivedCb: func() {
			numBlocksReceived.Inc()
		},
		OnDaSubmissionFailedCb: func() {
			numDaSubmissionsFailed.Inc()
		},
		OnDaSubmittedCb: func(duration time.Duration) {
			submissionDuration.Observe(float64(duration.Milliseconds()))
		},
		OnRetriesRequiredCb: func(retries int) {
			retriesHistogram.Observe(float64(retries))
		},
		OnInvalidNonceCb: func() {
			numInvalidNonces.Inc()
		},
	}, nil
}

func startMetrics(metricsAddr string, reg prometheus.Gatherer) (<-chan error, func(ctx context.Context) error) {
	errC := make(chan error, 1)
	server := &http.Server{Addr: metricsAddr, Handler: nil}

	http.Handle("/metrics", promhttp.HandlerFor(reg, promhttp.HandlerOpts{}))

	shutdown := func(ctx context.Context) error {
		if err := server.Shutdown(ctx); err != nil {
			return err
		}

		return nil
	}

	go func() {
		if err := server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
			errC <- err
		}
	}()

	return errC, shutdown
}

func StartMetricsServer(ctx context.Context, metricsAddr string, reg prometheus.Gatherer, logger sdklogging.Logger) {
	const (
		RetryCount    = 3
		RetryInterval = time.Second
	)

	errC, shutdownMetrics := startMetrics(metricsAddr, reg)
	go func() {
		retryCount := 0
		for {
			select {
			case err := <-errC:
				if err == nil {
					continue
				}

				if retryCount >= RetryCount {
					logger.Error("Failed to restart metrics server after multiple attempts", "err", err)
					return
				} else {
					logger.Error("Metrics server error", "err", err)
				}

				err = shutdownMetrics(ctx)
				if err != nil {
					logger.Error("Error while shutting down", "err", err)
					return
				}

				// Sleep before restart
				time.Sleep(RetryInterval)

				logger.Info("Attempting to restart metrics server")
				errC, shutdownMetrics = startMetrics(metricsAddr, reg)
				retryCount++
			case <-ctx.Done():
				return
			}
		}
	}()
}

File: relayer-rs/src/metrics.rs
use prometheus::{Counter, Encoder, Histogram, HistogramOpts, IntCounter, Registry};
use std::sync::Arc;
use tokio::time::Duration;
use warp::Filter;
use anyhow::Result;

const RELAYER_NAMESPACE: &str = "sffl_relayer";

pub struct RelayerMetrics {
    pub num_blocks_received: IntCounter,
    pub num_da_submissions_failed: IntCounter,
    pub submission_duration_ms: Histogram,
    pub retries_histogram: Histogram,
    pub num_of_invalid_nonces: IntCounter,
    pub num_of_expired_txs: IntCounter,
    pub num_of_timeout_txs: IntCounter,
}

impl RelayerMetrics {
    pub fn new(registry: &Registry) -> Result<Arc<Self>> {
        let num_blocks_received = IntCounter::new(
            format!("{}_num_blocks_received", RELAYER_NAMESPACE),
            "The number of blocks received from rollup",
        )?;
        registry.register(Box::new(num_blocks_received.clone()))?;

        let num_da_submissions_failed = IntCounter::new(
            format!("{}_num_da_submissions_failed", RELAYER_NAMESPACE),
            "The number of failed da submissions",
        )?;
        registry.register(Box::new(num_da_submissions_failed.clone()))?;

        let latency_buckets = vec![
            1.0, 25.0, 50.0, 75.0, 100.0, 250.0, 500.0, 1000.0, 2000.0, 3000.0, 4000.0, 5000.0,
            10000.0, f64::INFINITY,
        ];

        let submission_duration_ms = Histogram::with_opts(
            HistogramOpts::new(
                format!("{}_submission_duration_ms", RELAYER_NAMESPACE),
                "Duration of successful DA submissions",
            )
            .buckets(latency_buckets),
        )?;
        registry.register(Box::new(submission_duration_ms.clone()))?;

        let retries_histogram = Histogram::with_opts(
            HistogramOpts::new(
                format!("{}_retries_histogram", RELAYER_NAMESPACE),
                "Histogram of retry counts",
            )
            .buckets(prometheus::linear_buckets(0.0, 1.0, 3).unwrap()),
        )?;
        registry.register(Box::new(retries_histogram.clone()))?;

        let num_of_invalid_nonces = IntCounter::new(
            format!("{}_num_of_invalid_nonces", RELAYER_NAMESPACE),
            "Number of InvalidNonce error",
        )?;
        registry.register(Box::new(num_of_invalid_nonces.clone()))?;

        let num_of_expired_txs = IntCounter::new(
            format!("{}_num_of_expired_txs", RELAYER_NAMESPACE),
            "Number of Expired transactions",
        )?;
        registry.register(Box::new(num_of_expired_txs.clone()))?;

        let num_of_timeout_txs = IntCounter::new(
            format!("{}_num_of_timeout_txs", RELAYER_NAMESPACE),
            "Number of Timeout transactions",
        )?;
        registry.register(Box::new(num_of_timeout_txs.clone()))?;

        Ok(Arc::new(Self {
            num_blocks_received,
            num_da_submissions_failed,
            submission_duration_ms,
            retries_histogram,
            num_of_invalid_nonces,
            num_of_expired_txs,
            num_of_timeout_txs,
        }))
    }
}

pub async fn start_metrics_server(metrics_addr: String, registry: Registry) {
    let metrics_route = warp::path!("metrics").and(warp::get()).map(move || {
        let encoder = prometheus::TextEncoder::new();
        let mut buffer = Vec::new();
        encoder.encode(&registry.gather(), &mut buffer).unwrap();
        String::from_utf8(buffer).unwrap()
    });

    let socket_addr: std::net::SocketAddr = metrics_addr.parse().expect("Failed to parse metrics address");
    warp::serve(metrics_route).run(socket_addr).await;
}

pub trait EventListener {
    fn on_block_received(&self);
    fn on_da_submission_failed(&self);
    fn on_da_submitted(&self, duration: Duration);
    fn on_retries_required(&self, retries: i32);
    fn on_invalid_nonce(&self);
    fn on_expired_tx(&self);
    fn on_timeout_tx(&self);
}

impl EventListener for Arc<RelayerMetrics> {
    fn on_block_received(&self) {
        self.num_blocks_received.inc();
    }

    fn on_da_submission_failed(&self) {
        self.num_da_submissions_failed.inc();
    }

    fn on_da_submitted(&self, duration: Duration) {
        self.submission_duration_ms.observe(duration.as_millis() as f64);
    }

    fn on_retries_required(&self, retries: i32) {
        self.retries_histogram.observe(retries as f64);
    }

    fn on_invalid_nonce(&self) {
        self.num_of_invalid_nonces.inc();
    }

    fn on_expired_tx(&self) {
        self.num_of_expired_txs.inc();
    }

    fn on_timeout_tx(&self) {
        self.num_of_timeout_txs.inc();
    }
}